curvefs/src/mds/topology/topology_metric.cpp

```diff
@@ -0,0 +1,204 @@
+        CalcMetaServerMetrics(copysets, &metaServerMetricInfo);
+
+        auto it = gPoolMetrics.find(pid);
+        if (it == gPoolMetrics.end()) {
+            PoolMetricPtr lptr(new PoolMetric(poolName));
+            it = gPoolMetrics.emplace(pid, std::move(lptr)).first;
+        }
+
+        it->second->metaServerNum.set_value(metaServerMetricInfo.size());
+        it->second->copysetNum.set_value(copysets.size());
+
+        uint64_t totalInodeNum = 0;
+        uint64_t totalDentryNum = 0;
+        std::list<Partition> partitions = topo_->GetPartitionInfosInPool(pid);
+        for (auto pit = partitions.begin(); pit != partitions.end(); ++pit) {
+            totalInodeNum += pit->GetInodeNum();
+            totalDentryNum += pit->GetDentryNum();
+        }
+        it->second->inodeNum.set_value(totalInodeNum);
+        it->second->dentryNum.set_value(totalDentryNum);
+        it->second->partitionNum.set_value(partitions.size());
+
+
+        uint64_t totalDiskCapacity = 0;
+        uint64_t totalDiskUsed = 0;
+
+        // process the metric of metaserver.
+        for (auto cm : metaServerMetricInfo) {
+            auto ix = gMetaServerMetrics.find(cm.first);
+            if (ix == gMetaServerMetrics.end()) {
+                MetaServerMetricPtr cptr(new MetaServerMetric(cm.first));
+                ix =
+                    gMetaServerMetrics.emplace(cm.first, std::move(cptr)).first;
+            }
+            ix->second->scatterWidth.set_value(cm.second.scatterWidth);
+            ix->second->copysetNum.set_value(cm.second.copysetNum);
+            ix->second->leaderNum.set_value(cm.second.leaderNum);
+            ix->second->partitionNum.set_value(cm.second.partitionNum);
+
+            totalDiskCapacity += ix->second->diskCapacity.get_value();
+            totalDiskUsed += ix->second->diskUsed.get_value();
+        }
+
+        it->second->diskCapacity.set_value(totalDiskCapacity);
+        it->second->diskUsed.set_value(totalDiskUsed);
+    }
+
+    // remove pool metrics that no longer exist
+    for (auto it = gPoolMetrics.begin(); it != gPoolMetrics.end();) {
+        if (std::find(pools.begin(), pools.end(), it->first) == pools.end()) {
+            it = gPoolMetrics.erase(it);
+        } else {
+            it++;
+        }
+    }
+
+    // remove metaservers that no longer exist
+    for (auto it = gMetaServerMetrics.begin();
+         it != gMetaServerMetrics.end();) {
+        if (std::find(metaservers.begin(), metaservers.end(), it->first) ==
+            metaservers.end()) {
+            it = gMetaServerMetrics.erase(it);
+        } else {
+            it++;
+        }
+    }
+
+    return;
+}
+
+void TopologyMetricService::CalcMetaServerMetrics(
+    const std::vector<CopySetInfo> &copysets,
+    std::map<MetaServerIdType, MetaServerMetricInfo> *msMetricInfoMap) {
+    for (auto &cs : copysets) {
+        for (auto &msId : cs.GetCopySetMembers()) {
```

Comment:
for (const auto &cs : copysets)
    for (const auto &msId : cs.GetCopySetMembers())

Reply:
done